### PR TITLE
fix(Fela): prevent styles crash on invalid CSS property provided

### PR DESCRIPTION
**Pull request by kuzhelov**
# Problem

### TL;DR
When, for some reason, for CSS property provided a value that ends with opening bracket (or, in general, a value with invalid braces sequence), this results in shadowing all the subsequent Fela styles added to the Fela's rules `<style />` DOM element.

This problem shows up quite often while playing with docs, as it is typical situation when not fully-provided variable values are applied.

---------------

Specifically, suppose that we have the following style object defined for our component:

```jsx
const style = (props) => {
   ...
   color: 'rgba('
}
```

When component is rendered, its styles are transformed into class declarations and added to Fela's `<style>` DOM element on the page:

```html
<style>
   .a{ display: block; color: rgb( } .b{ display: inline; color: red } ...
</ style>
```

The problem is that invalid sequence of brackets (that was provided as part of the component's CSS property value) results in browser engine ignoring the rest of the `<style />` element content - so that, essentially, all subsequent classes are like not declared.

```html
<style>
    .a{ display: block; color: rgb(       // everything that goes next is ignored by browser
</ style>
```

This results in absence of styling for components that were relying on these classes:

```html
<!-- oops, styles WON'T be applied -->
<div class="b">
   ...
</div>
```

# Solution

Proposed solution is to introduce plugin that will be responsible for ignoring CSS properties that have improper braces sequence as their value. We can extend this functionality to handle other cases if necessary - till we won't find a better solution (problem will be fixed on the Fela's side, new widely-used plugin will be introduced for handling these cases, etc).

### CODE REVIEW

File: test/specs/lib/felaSanitizeCssPlugin-test.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        ...validStringStyle,
+        ...invalidCssStyle,
+      },
+    }
+
+    expect(sanitize(style)).toEqual({
+      ...validStringStyle,
+      '::before': {
+        ...validStringStyle,
+      },
```
Comment (levithomason) [inline test/specs/lib/felaSanitizeCssPlugin-test.ts]: To the last point, I think having simple objects here would help understanding.  It is not clear what the test is doing from looking at it.  I need to scroll back and forth, memorize two object shapes, and then picture two different merged shapes in my mind to see the assertion.

I think having the objects in simple plain form is more ideal for testing purposes.

Feedback?

File: test/specs/lib/felaSanitizeCssPlugin-test.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      skip: [cssPropertyToExclude],
+    })
+
+    const style = {
+      validStringStyle,
+      [cssPropertyToExclude]: invalidCssValue,
+    }
+
+    expect(sanitize(style)).toEqual(style)
+  })
```
Comment (levithomason) [inline test/specs/lib/felaSanitizeCssPlugin-test.ts]: Would encourage less abstraction in testing.  Consider this comparison:

```jsx
const validStringStyle = {
  display: 'block',
  margin: '0 0 0 0',
}

const invalidCssValue = 'rgba('

test('should skip excluded CSS props', () => {
  const cssPropertyToExclude = 'myProperty'

  const sanitize = sanitizeCss({
    skip: [cssPropertyToExclude],
  })

  const style = {
    validStringStyle,
    [cssPropertyToExclude]: invalidCssValue,
  }

  expect(sanitize(style)).toEqual(style)
})
```

Same without abstraction:

```jsx
test('should skip excluded CSS props', () => {
  const sanitize = sanitizeCss({ skip: ['myProperty'] })

  const style = {
    validStringStyle: { display: 'block', margin: '0 0 0 0' },
    myProperty: 'rgba(',
  }

  expect(sanitize(style)).toEqual(style)
})
```

One or two non-dry properties are OK if the trade off is readability.  Suggest always prioritizing "code is for humans" above the most optimal solution.

File: src/lib/felaSanitizeCssPlugin.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  if (typeof value !== 'string') {
+    return true
+  }
+
+  const openingBrackets = '({['
+  const closingBrackets = ')}]'
+
+  const openingBracketsStack = []
+
+  for (let i = 0; i < value.length; ++i) {
```
Comment (levithomason) [inline src/lib/felaSanitizeCssPlugin.ts]: Could we get an inline comment here as to what this is trying to achieve?

File: test/specs/lib/felaSanitizeCssPlugin-test.ts (hunk 1 above)
Comment (kuzhelov) [reply test/specs/lib/felaSanitizeCssPlugin-test.ts]: sure, will do this way, agree

File: test/specs/lib/felaSanitizeCssPlugin-test.ts (hunk 2 above)
Comment (kuzhelov) [reply test/specs/lib/felaSanitizeCssPlugin-test.ts]: yes, totally valid point, will rework

File: src/lib/felaSanitizeCssPlugin.ts (hunk 3 above)
Comment (kuzhelov) [reply src/lib/felaSanitizeCssPlugin.ts]: sure, will add this
